docs/discovery-naming-conventions.md

```diff
@@ -0,0 +1,137 @@
+Endpoints is called the _Discovered Name_, and is built from the following
+_Components_:
+
+```
+${discoverer-prefix}-${service-name}-${backend-name}
+```
+
+The name of service ports is not specified, and is handled independently by each
+discoverer implementation.
+
+## Kubernetes Service Naming Requirements
+
+Kubernetes Service names must adhere to the [rfc1035 DNS Label](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/identifiers.md) specification:
+
+> An alphanumeric (a-z, and 0-9) string, with a maximum length of 63 characters,
+> with the '-' character allowed anywhere except the first or last character,
+> suitable for use as a hostname or segment in a domain name.
+
+### Handling of invalid characters
+
+Backend systems other than Kubernetes (e.g. OpenStack) might support names that
+contain characters that are invalid according to the Kubernetes Service naming
+requirements. In these scenarios, discoverers will skip the service and inform
+the user via logs and metrics that the backend service name is not compatible
+with Gimbal.
+
+Even thought this might affect the user experience and cause friction, Gimbal
+prefers to be predictable and safe when it comes to managing backend services
+and endpoints, instead of using "magic" (replacing or removing characters, etc)
+to massage invalid names into valid ones.
+
+### Handling Discovered Names that are longer than 63 characters
+
+When the _Discovered Name_ is longer than 63 characters, it is shortened using
+the following process:
+
+1. Each component of the _Discovered Name_ gets the same number of maximum
+   characters allowed. The _Discovered Name_ has a total of three components,
+   and thus each component is allocated a maximum of 21 characters.
+
+2. Take the SHA256 hash of the _Discovered Name_ before shortening, and take the
+   first 6 characters of the resulting hash. This is called the _short hash_.
+
+3. Starting at the last _Discovered Name Component_, check whether it is longer
+   than the allocated number of characters. If it is, truncate the end of the
+   _Component_ and append `-${short-hash}`.
+
+4. If the resulting _Discovered Name_ is still longer than 63 characters, move
+   onto the next _Component_ and shorten using the _short hash_. Continue until
+   the resulting _Discovered Name_ is 63 characters or less.
+
+5. In the case that the _Discovered Name_ is longer than 63 characters after the
+   shortening process, the SHA256 hash produced in step 2 will be used as the
+   _Discovered Name_, prepended with `g-` to ensure the name starts with a
+   letter. This scenario will only happen if *all* _Components_ of the initial
+   _Discovered Name_ are longer than 21 characters, which is a rare case.
+
+#### Example
+
+- `${discoverer-prefix}`: `k8s`
+- `${service-name}`: `the-really-long-kube-service-name-that-is-exactly-63-characters`
+- `${backend-name}`: `us-east-cluster`
```

Comment:
Should we impose a limit on the backend-name length?

Reply:
Hmmm. Any reason for doing it? Wondering how we would come up with that limit.

Reply:
Just to help keep the name field bigger. So instead of 21 characters for cluster-name you could make it 10, then the name field becomes 32 chars. Not sure how often we run into this though.

Reply:
Got it. The downside is that cluster names are now smaller, so that could also be a problem. I agree that we should probably try to find a sweet spot here though.

Reply:
I think until we know more about real-world use-cases, we should split them equally.